include/ensmallen_bits/moead/moead_impl.hpp

```diff
@@ -0,0 +1,453 @@
+        + std::to_string(neighborSize) + " was detected."
+    );
+  }
+
+  if (neighborSize > populationSize - 1u)
+  {
+    std::ostringstream oss;
+    oss << "MOEAD::Optimize(): " << "neighborSize is " << neighborSize
+        << " but populationSize is " << populationSize << "(should be"
+        << " atleast " << (neighborSize + 1u) << ")" << std::endl;
+    throw std::logic_error(oss.str());
+  }
+
+  if (distributionIndex < 0)
+  {
+    throw std::invalid_argument(
+        "distributionIndex should be positive, however "
+        + std::to_string(distributionIndex) + " was detected.");
+  }
+
+  if (differentialWeight > 1.0 || differentialWeight < 0.0)
+  {
+    throw std::invalid_argument(
+        "The parameter differentialWeight needs to be in [0,1], however "
+        + std::to_string(differentialWeight) + " was detected.");
+  }
+
+  if (neighborProb > 1.0 || neighborProb < 0.0)
+  {
+    throw std::invalid_argument(
+        "The parameter neighborProb needs to be in [0,1], however "
+        + std::to_string(neighborProb) + " was detected.");
+  }
+
+  // Check if lower bound is a vector of a single dimension.
+  if (lowerBound.n_rows == 1)
+    lowerBound = lowerBound(0, 0) * arma::ones(iterate.n_rows, iterate.n_cols);
+
+  // Check if lower bound is a vector of a single dimension.
+  if (upperBound.n_rows == 1)
+    upperBound = upperBound(0, 0) * arma::ones(iterate.n_rows, iterate.n_cols);
+
+  // Convenience typedefs.
+  typedef typename MatType::elem_type ElemType;
+
+  // Number of objective functions. Represented by M in the paper.
+  numObjectives = sizeof...(ArbitraryFunctionType);
+  // Dimensionality of variable space. Also referred to as number of genes.
+  size_t numVariables = iterate.n_rows;
+
+  if (numObjectives < 2u)
+  {
+    throw std::logic_error("MOEAD::Optimize(): This is a multiobjective problem, "
+        "numObjectives must be atleast 2.");
+  }
+
+  if (lowerBound.size() != upperBound.size())
+  {
+    throw std::logic_error("MOEAD::Optimize(): size of lowerBound and upperBound "
+        "must be equal.");
+  }
+
+  if (lowerBound.size() != iterate.n_elem)
+  {
+    throw std::logic_error("MOEAD::Optimize(): there should be a lower bound and "
+        "an upper bound for each variable in the initial point.");
+  }
+
+  for (size_t geneIdx = 0; geneIdx < lowerBound.size(); geneIdx++)
+  {
+    if (lowerBound[geneIdx] >= upperBound[geneIdx])
+    {
+      std::ostringstream ss;
+      ss << "MOEAD::Optimize():" << "the lowerBound value: " << lowerBound[geneIdx]
+         << " is greater than upperBound value: " << upperBound[geneIdx]
+         << " at index: " << geneIdx << std::endl;
+
+      throw std::logic_error(ss.str());
+    }
+  }
+
+  bool terminate = false;
+
+  arma::uvec shuffle;
+  // The weight matrix. Each vector represents a decomposition subproblem.
+  arma::Mat<ElemType> weights(numObjectives, populationSize, arma::fill::randu);
+  weights += 1E-10; // Numerical stability
```

Comment:
Let us put that in the constructor, see https://github.com/mlpack/ensmallen/blob/master/include/ensmallen_bits/adam/adam.hpp#L98 for an example.

Reply:
I guess, still a good idea to add `epsilon` into the constructor and use that parameter here.